File: verp_staffing/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py
# Copyright (c) 2026, Vrugle and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, get_last_day
import datetime


def execute(filters=None):
    filters = frappe._dict(filters or {})

    # ── Server-side defaults (identical to Balance Sheet) ─────────
    if not filters.company:
        filters.company = frappe.db.get_single_value("Accounts Settings", "default_company")

    if not filters.filter_based_on:
        filters.filter_based_on = "Fiscal Year"

    if filters.filter_based_on == "Fiscal Year":
        if not filters.from_fiscal_year or not filters.to_fiscal_year:
            latest_fy = _get_latest_fiscal_year(filters.company)
            if latest_fy:
                if not filters.from_fiscal_year:
                    filters.from_fiscal_year = latest_fy
                if not filters.to_fiscal_year:
                    filters.to_fiscal_year = latest_fy

    validate_filters(filters)

    columns     = get_columns(filters)
    data, chart = get_data(filters)

    return columns, data, None, chart


# ─────────────────────────────────────────────
# AUTO-RESOLVE LATEST FISCAL YEAR FOR COMPANY
# ─────────────────────────────────────────────

def _get_latest_fiscal_year(company):
    if not company:
        return None
    rows = frappe.db.sql("""
        SELECT fy.name
        FROM `tabFiscal Year` fy
        INNER JOIN `tabFiscal Year Company` fyc ON fyc.parent = fy.name
        WHERE fyc.company = %s
        ORDER BY fy.year_start_date DESC
        LIMIT 1
    """, company, as_dict=True)
    return rows[0].name if rows else None


# ─────────────────────────────────────────────
# VALIDATION
# ─────────────────────────────────────────────

def validate_filters(filters):
    if not filters.company:
        frappe.throw(_("Please select a Company (or set a Default Company in Accounts Settings)."))

    if filters.filter_based_on == "Fiscal Year":
        if not filters.from_fiscal_year or not filters.to_fiscal_year:
            frappe.throw(_(
                "Could not determine a Fiscal Year for company '{0}'. "
                "Please ensure the Fiscal Year has this company in Included Companies."
            ).format(filters.company))
    else:
        if not filters.from_date or not filters.to_date:
            frappe.throw(_("Please select From Date and To Date."))
        if getdate(filters.from_date) > getdate(filters.to_date):
            frappe.throw(_("From Date cannot be greater than To Date."))


# ─────────────────────────────────────────────
# DATE HELPERS  (identical to Balance Sheet)
# ─────────────────────────────────────────────

def get_from_to_dates(filters):
    if filters.filter_based_on == "Fiscal Year":
        from_fy = frappe.get_doc("Fiscal Year", filters.from_fiscal_year)
        to_fy   = frappe.get_doc("Fiscal Year", filters.to_fiscal_year)
        return getdate(from_fy.year_start_date), getdate(to_fy.year_end_date)
    else:
        return getdate(filters.from_date), getdate(filters.to_date)


def get_period_date_ranges(filters):
    from_date, to_date = get_from_to_dates(filters)
    periodicity = filters.get("periodicity", "Yearly")

    if periodicity == "Yearly":
        label = "{0} to {1}".format(
            frappe.format(from_date, {"fieldtype": "Date"}),
            frappe.format(to_date,   {"fieldtype": "Date"})
        )
        return [(label, from_date, to_date)]

    period_list = []
    start = from_date
    end   = to_date

    while start <= end:
        if periodicity == "Monthly":
            period_end = get_last_day(start)
            label      = start.strftime("%b %Y")

        elif periodicity == "Quarterly":
            quarter_month_end = ((start.month - 1) // 3 + 1) * 3
            period_end = get_last_day(datetime.date(start.year, quarter_month_end, 1))
            label      = "Q{0} {1}".format(((start.month - 1) // 3) + 1, start.year)

        elif periodicity == "Half-Yearly":
            if start.month <= 6:
                period_end = datetime.date(start.year, 6, 30)
                label      = "H1 {0}".format(start.year)
            else:
                period_end = datetime.date(start.year, 12, 31)
                label      = "H2 {0}".format(start.year)
        else:
            period_end = end
            label      = "Period"

        actual_end = min(getdate(period_end), end)
        period_list.append((label, start, actual_end))
        start = getdate(period_end) + datetime.timedelta(days=1)

    return period_list


# ─────────────────────────────────────────────
# COLUMNS
# ─────────────────────────────────────────────

def get_columns(filters):
    columns = [{
        "label":     _("Account"),
        "fieldname": "account",
        "fieldtype": "Link",
        "options":   "Account",
        "width":     300
    }]

    view        = filters.get("view", "Report View")
    period_list = get_period_date_ranges(filters)

    for label, _from, _to in period_list:
        fn = frappe.scrub(label)

        if view == "Margin View":
            # Margin View: one Percent column per period (% of Total Income)
            # The actual value stored in data[fn + "_margin"] by Python
            # We reuse the same fieldname so the formatter can intercept it
            columns.append({
                "label":     _("{0} Margin %".format(label)),
                "fieldname": fn,
                "fieldtype": "Percent",
                "width":     160
            })
        else:
            # Report View / Growth View: currency column
            columns.append({
                "label":     _(label),
                "fieldname": fn,
                "fieldtype": "Currency",
                "options":   "currency",
                "width":     160
            })
            # Growth View: additional % change column after first period
            if view == "Growth View" and len(period_list) > 1:
                columns.append({
                    "label":     _("{0} Growth %".format(label)),
                    "fieldname": fn + "_growth",
                    "fieldtype": "Percent",
                    "width":     120
                })

    return columns


# ─────────────────────────────────────────────
# ACCOUNT TREE  — P&L uses Income + Expense only
# ─────────────────────────────────────────────

def get_accounts(company):
    """
    Fetch Income and Expense accounts for this company.
    root_type IN ('Income', 'Expense') — this is the key difference from Balance Sheet.
    """
    return frappe.db.sql("""
        SELECT name, account_name, parent_account,
               root_type, account_type, lft, rgt, is_group
        FROM `tabAccount`
        WHERE company   = %s
          AND root_type IN ('Income', 'Expense')
        ORDER BY lft
    """, company, as_dict=True)


# ─────────────────────────────────────────────
# GL BALANCE QUERY  (identical to Balance Sheet)
# ─────────────────────────────────────────────

def get_gl_balances(filters, from_date, to_date, account_names):
    if not account_names:
        return {}

    placeholders = ", ".join(["%s"] * len(account_names))
    conditions   = [
        "gle.company = %s",
        "gle.account IN ({0})".format(placeholders),
        "gle.posting_date BETWEEN %s AND %s",
        "gle.is_cancelled = 0"
    ]
    values = [filters.company] + list(account_names) + [from_date, to_date]

    if filters.get("finance_book"):
        if filters.get("include_default_fb_entries"):
            conditions.append(
                "(gle.finance_book = %s OR gle.finance_book IS NULL OR gle.finance_book = '')"
            )
        else:
            conditions.append("gle.finance_book = %s")
        values.append(filters.finance_book)

    rows = frappe.db.sql("""
        SELECT gle.account,
               SUM(gle.debit)  AS debit,
               SUM(gle.credit) AS credit
        FROM `tabGL Entry` gle
        WHERE {cond}
        GROUP BY gle.account
    """.format(cond=" AND ".join(conditions)), values, as_dict=True)

    return {r.account: r for r in rows}


# ─────────────────────────────────────────────
# NET BALANCE — sign convention for P&L
#
# Income accounts: credit-normal → net = credit - debit (positive = income earned)
# Expense accounts: debit-normal → net = debit - credit (positive = expense incurred)
#
# This makes every row display as a positive number when the account is
# used normally, which is the standard P&L presentation.
# ─────────────────────────────────────────────

def compute_net(account, gl_map, root_type):
    row    = gl_map.get(account, {})
    debit  = flt(row.get("debit",  0))
    credit = flt(row.get("credit", 0))

    if root_type == "Income":
        return credit - debit   # income is credit-normal
    else:
        return debit - credit   # expense is debit-normal


def get_group_total(account_name, gl_map, children_map, acc_map):
    acc   = acc_map.get(account_name)
    total = compute_net(account_name, gl_map, acc.root_type if acc else "Expense")
    for child in children_map.get(account_name, []):
        total += get_group_total(child, gl_map, children_map, acc_map)
    return total


# ─────────────────────────────────────────────
# MAIN DATA BUILDER
# ─────────────────────────────────────────────

def get_data(filters):
    accounts    = get_accounts(filters.company)
    period_list = get_period_date_ranges(filters)
    from_date, _to_date = get_from_to_dates(filters)

    # ── P&L accumulation logic ────────────────────────────────────
    # Default (unchecked): each period shows ONLY that period's movement — natural for P&L.
    # show_accumulated_values checked: each period shows YTD cumulative totals.
    accumulated = bool(filters.get("show_accumulated_values", 0))

    show_zero = bool(filters.get("show_zero_values", 0))
    view      = filters.get("view", "Report View")

    # ── Lookup structures ─────────────────────────────────────────
    acc_map      = {a.name: a for a in accounts}
    children_map = {}
    for a in accounts:
        children_map.setdefault(a.parent_account, []).append(a.name)

    all_account_names = [a.name for a in accounts]

    # ── Fetch GL per period ───────────────────────────────────────
    period_gl = []
    for label, p_from, p_to in period_list:
        query_from = from_date if accumulated else p_from
        gl = get_gl_balances(filters, query_from, p_to, all_account_names)
        period_gl.append((label, p_from, p_to, gl))

    # ── Row builders ──────────────────────────────────────────────
    data = []

    def section_header(label, root_type):
        data.append({
            "account":   label,
            "root_type": root_type,
            "indent":    0,
            "is_group":  1,
            "bold":      1,
            "currency":  filters.get("currency"),
        })

    def add_rows(account_name, indent=1):
        acc = acc_map.get(account_name)
        if not acc:
            return

        row = {
            "account":   acc.account_name,
            "root_type": acc.root_type,
            "indent":    indent,
            "is_group":  acc.is_group,
            "currency":  filters.get("currency"),
        }

        has_value = False
        prev_val  = None

        for label, _pf, _pt, gl in period_gl:
            fn = frappe.scrub(label)

            net = (get_group_total(account_name, gl, children_map, acc_map)
                   if acc.is_group
                   else compute_net(account_name, gl, acc.root_type))

            row[fn] = net

            if net:
                has_value = True

            if view == "Growth View" and len(period_list) > 1 and prev_val is not None:
                growth = ((net - prev_val) / abs(prev_val) * 100) if prev_val else 0.0
                row[fn + "_growth"] = growth

            prev_val = net

        # Suppress zero leaf rows unless show_zero is on
        if not show_zero and not has_value and not acc.is_group:
            return

        data.append(row)

        for child in children_map.get(account_name, []):
            add_rows(child, indent + 1)

    def section_total_row(label, root_type):
        """Sum all leaf accounts of the given root_type."""
        row = {
            "account":  label,
            "indent":   0,
            "is_group": 1,
            "bold":     1,
            "currency": filters.get("currency"),
        }
        for lbl, _pf, _pt, gl in period_gl:
            fn = frappe.scrub(lbl)
            row[fn] = sum(
                compute_net(a.name, gl, a.root_type)
                for a in accounts
                if a.root_type == root_type and not a.is_group
            )
        return row

    # ── INCOME ────────────────────────────────────────────────────
    section_header("INCOME", "Income")
    for root in [a.name for a in accounts
                 if a.root_type == "Income" and not a.parent_account]:
        add_rows(root, indent=1)
    income_total = section_total_row("Total Income", "Income")
    data.append(income_total)
    data.append({})  # spacer

    # ── EXPENSES ──────────────────────────────────────────────────
    section_header("EXPENSES", "Expense")
    for root in [a.name for a in accounts
                 if a.root_type == "Expense" and not a.parent_account]:
        add_rows(root, indent=1)
    expense_total = section_total_row("Total Expenses", "Expense")
    data.append(expense_total)
    data.append({})  # spacer

    # ── NET PROFIT / LOSS ─────────────────────────────────────────
    # Net Profit = Total Income − Total Expenses
    # Positive → Profit (green in formatter), Negative → Loss (red)
    net_row = {
        "account":        "Net Profit / Loss",
        "indent":         0,
        "is_group":       1,
        "bold":           1,
        "is_net_profit":  1,   # picked up by JS formatter for green/red colouring
        "currency":       filters.get("currency"),
    }
    for label, _pf, _pt in period_list:
        fn = frappe.scrub(label)
        net_row[fn] = flt(income_total.get(fn, 0)) - flt(expense_total.get(fn, 0))

    data.append(net_row)

    # ── MARGIN VIEW post-pass ─────────────────────────────────────
    # After all rows are built, compute each row's value as a % of
    # Total Income for that period and store as fn + "_margin".
    # The JS formatter reads this key when view == "Margin View".
    if view == "Margin View":
        for row in data:
            if not row:
                continue  # spacer rows — skip
            for label, _pf, _pt in period_list:
                fn           = frappe.scrub(label)
                total_income = flt(income_total.get(fn, 0))
                row_val      = flt(row.get(fn, 0))
                row[fn + "_margin"] = (
                    round((row_val / total_income) * 100, 2) if total_income else 0.0
                )

    chart = get_chart_data(data, period_list)
    return data, chart


# ─────────────────────────────────────────────
# CHART
# ─────────────────────────────────────────────

def get_chart_data(data, period_list):
    labels        = [p[0] for p in period_list]
    income_row    = next((d for d in data if d.get("account") == "Total Income"),    {})
    expense_row   = next((d for d in data if d.get("account") == "Total Expenses"),  {})
    net_row       = next((d for d in data if d.get("account") == "Net Profit / Loss"), {})

    datasets = []
    for name, row, color in [
        ("Income",          income_row,  "#5ef4a0"),
        ("Expenses",        expense_row, "#f4825e"),
        ("Net Profit/Loss", net_row,     "#5e81f4"),
    ]:
        datasets.append({
            "name":      name,
            "values":    [flt(row.get(frappe.scrub(lbl), 0)) for lbl in labels],
            "chartType": "bar"
        })

    return {
        "data":   {"labels": labels, "datasets": datasets},
        "type":   "bar",
        "colors": ["#5ef4a0", "#f4825e", "#5e81f4"],
        "title":  "Profit and Loss Statement"
    }

File: verp_staffing/accounts/report/profit_and_loss_statement/profit_and_loss_statement.js
// Copyright (c) 2026, Vrugle and contributors
// For license information, please see license.txt

// profit_and_loss_statement.js
// Frappe Script Report – Profit and Loss Statement
// Identical filter behaviour to Balance Sheet report.

frappe.query_reports["Profit and Loss Statement"] = {
	// ─────────────────────────────────────────────────────────────
	// INTERNAL: restrict fiscal year dropdowns to company's years
	// ─────────────────────────────────────────────────────────────
	_set_fiscal_year_query: function (fy_names) {
		["from_fiscal_year", "to_fiscal_year"].forEach(function (fname) {
			let filter = frappe.query_report.get_filter(fname);
			if (!filter) return;
			filter.df.get_query = function () {
				return { filters: [["Fiscal Year", "name", "in", fy_names]] };
			};
			if (filter.df && filter.df.$input) {
				let ctrl = filter.df.$input.data("fieldobj");
				if (ctrl) ctrl.get_query = filter.df.get_query;
			}
		});
	},

	// ─────────────────────────────────────────────────────────────
	// INTERNAL: cascade company → currency, finance_book, fiscal years
	// force_fy = true  → always overwrite fiscal year values
	// force_fy = false → only overwrite if current value is invalid
	// ─────────────────────────────────────────────────────────────
	_apply_company_defaults: function (company, force_fy) {
		let me = frappe.query_reports["Profit and Loss Statement"];

		// 1. Company → default_currency
		frappe.db.get_value("Company", company, ["default_currency"], function (r) {
			if (!r) return;
			if (r.default_currency) {
				frappe.query_report.set_filter_value("currency", r.default_currency);
			}

			// ── NEW: reset exchange rate field when company changes ──────
			let selected = frappe.query_report.get_filter_value("currency");
			me._toggle_exchange_rate_field(selected, r.default_currency);
		});

		// 2. Fiscal years whose included_companies contains this company
		frappe.db
			.get_list("Fiscal Year", {
				fields: ["name"],
				filters: [["Fiscal Year Company", "company", "=", company]],
				order_by: "year_start_date asc",
				limit: 500,
			})
			.then(function (rows) {
				if (!rows || !rows.length) return;

				let fy_names = rows.map(function (r) {
					return r.name;
				});
				let last_fy = fy_names[fy_names.length - 1];

				me._set_fiscal_year_query(fy_names);

				if (force_fy) {
					frappe.query_report.set_filter_value("from_fiscal_year", last_fy);
					frappe.query_report.set_filter_value("to_fiscal_year", last_fy);
				} else {
					let cur_from = frappe.query_report.get_filter_value("from_fiscal_year");
					let cur_to = frappe.query_report.get_filter_value("to_fiscal_year");
					if (!cur_from || !fy_names.includes(cur_from)) {
						frappe.query_report.set_filter_value("from_fiscal_year", last_fy);
					}
					if (!cur_to || !fy_names.includes(cur_to)) {
						frappe.query_report.set_filter_value("to_fiscal_year", last_fy);
					}
				}
			});
	},

	// ─────────────────────────────────────────────────────────────
	// ONLOAD – auto-populate from Accounts Settings → Company
	// ─────────────────────────────────────────────────────────────
	onload: function (report) {
		let me = frappe.query_reports["Profit and Loss Statement"];

		frappe.db.get_value(
			"Accounts Settings",
			"Accounts Settings",
			"default_company",
			function (r) {
				let company =
					r && r.default_company
						? r.default_company
						: frappe.defaults.get_user_default("Company");

				if (!company) return;

				frappe.query_report.set_filter_value("company", company);
				frappe.query_report.set_filter_value("filter_based_on", "Fiscal Year");
				me._apply_company_defaults(company, true);
			},
		);
	},

	// ─────────────────────────────────────────────────────────────
	// FILTERS  (identical structure to Balance Sheet)
	// ─────────────────────────────────────────────────────────────
	filters: [
		// ── Row 1 ──────────────────────────────────────────────────
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			on_change: function () {
				let company = frappe.query_report.get_filter_value("company");
				if (company) {
					frappe.query_reports["Profit and Loss Statement"]._apply_company_defaults(
						company,
						true,
					);
				}
			},
		},
		{
			fieldname: "finance_book",
			label: __("Finance Book"),
			fieldtype: "Link",
			options: "Finance Book",
		},

		// ── Row 2 ──────────────────────────────────────────────────
		{
			fieldname: "filter_based_on",
			label: __("Filter Based On"),
			fieldtype: "Select",
			options: "Fiscal Year\nDate Range",
			default: "Fiscal Year",
			reqd: 1,
			on_change: function () {
				let val = frappe.query_report.get_filter_value("filter_based_on");
				frappe.query_report.toggle_filter_display(
					"from_fiscal_year",
					val !== "Fiscal Year",
				);
				frappe.query_report.toggle_filter_display("to_fiscal_year", val !== "Fiscal Year");
				frappe.query_report.toggle_filter_display("from_date", val !== "Date Range");
				frappe.query_report.toggle_filter_display("to_date", val !== "Date Range");
			},
		},

		// Fiscal Year sub-fields
		{
			fieldname: "from_fiscal_year",
			label: __("Start Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
			depends_on: "eval:doc.filter_based_on=='Fiscal Year'",
			reqd: 0,
		},
		{
			fieldname: "to_fiscal_year",
			label: __("End Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
			depends_on: "eval:doc.filter_based_on=='Fiscal Year'",
			reqd: 0,
		},

		// Date Range sub-fields
		{
			fieldname: "from_date",
			label: __("Start Date"),
			fieldtype: "Date",
			depends_on: "eval:doc.filter_based_on=='Date Range'",
			reqd: 0,
		},
		{
			fieldname: "to_date",
			label: __("End Date"),
			fieldtype: "Date",
			depends_on: "eval:doc.filter_based_on=='Date Range'",
			reqd: 0,
		},

		// ── Row 3 ──────────────────────────────────────────────────
		{
			fieldname: "periodicity",
			label: __("Periodicity"),
			fieldtype: "Select",
			options: "Yearly\nHalf-Yearly\nQuarterly\nMonthly",
			default: "Yearly",
			reqd: 1,
		},
		{
			fieldname: "currency",
			label: __("Currency"),
			fieldtype: "Link",
			options: "Currency",
			on_change: function () {
				let me = frappe.query_reports["Profit and Loss Statement"];
				let selected = frappe.query_report.get_filter_value("currency");
				let company = frappe.query_report.get_filter_value("company");
				if (!company) return;

				frappe.db.get_value("Company", company, "default_currency", function (r) {
					let company_currency = r && r.default_currency;
					me._toggle_exchange_rate_field(selected, company_currency);
				});
			},
		},
		{
			fieldname: "presentation_exchange_rate",
			label: __("Exchange Rate"),
			fieldtype: "Float",
			default: 1,
			hidden: 1,
			precision: 6,
			on_change: function () {
				let me = frappe.query_reports["Profit and Loss Statement"];
				let rate = frappe.query_report.get_filter_value("presentation_exchange_rate") || 1;
				let from = frappe.query_report.get_filter_value("currency");
				let company = frappe.query_report.get_filter_value("company");

				if (!from || !company) return;

				frappe.query_report.refresh();
			},
		},
		{
			fieldname: "view",
			label: __("Select View"),
			fieldtype: "Select",
			options: "Report View\nGrowth View\nMargin View",
			default: "Report View",
		},

		// ── Checkboxes ─────────────────────────────────────────────
		{
			fieldname: "show_accumulated_values",
			label: __("Show Accumulated Values (YTD)"),
			fieldtype: "Check",
			default: 0,
			description: __(
				"When checked, each period column shows year-to-date totals instead of that period's movement.",
			),
		},
		{
			fieldname: "include_default_fb_entries",
			label: __("Include Default FB Entries"),
			fieldtype: "Check",
			default: 1, // ON by default
		},
		{
			fieldname: "show_zero_values",
			label: __("Show Zero Values"),
			fieldtype: "Check",
			default: 0, // OFF by default
		},
	],

	// ─────────────────────────────────────────────────────────────
	// FORMATTER
	// ─────────────────────────────────────────────────────────────
	formatter: function (value, row, column, data, default_formatter) {
		if (!data) return value;

		let view = frappe.query_report.get_filter_value("view") || "Report View";

		// ── Margin View: all numeric columns show % of Total Income ─
		if (view === "Margin View" && column.fieldname !== "account") {
			// Python sets a  _margin  suffixed key for each period fieldname
			let margin_key = column.fieldname + "_margin";
			let margin = data[margin_key];

			if (margin === undefined || margin === null) {
				// Section headers / spacer rows have no margin — render blank
				return "";
			}

			let pct = frappe.utils.flt(margin, 2);
			let color =
				pct < 0 ? "var(--red-500)" : pct > 0 ? "var(--green-500)" : "var(--text-muted)";
			let bold = data.bold ? "font-weight:700;" : "";
			return `<span style="color:${color};${bold}">${pct}%</span>`;
		}

		// ── Growth View: Python sets _growth suffixed key ───────────
		if (
			view === "Growth View" &&
			column.fieldname !== "account" &&
			column.fieldname.endsWith("_growth")
		) {
			let pct = frappe.utils.flt(value, 2);
			let color = pct < 0 ? "var(--red-500)" : "var(--green-500)";
			return `<span style="color:${color}">${pct}%</span>`;
		}

		value = default_formatter(value, row, column, data);

		// Section headers: INCOME / EXPENSES
		if (data.indent === 0 && data.is_group && column.fieldname === "account") {
			value = `<span style="font-size:12px;font-weight:700;letter-spacing:0.08em;
				text-transform:uppercase;color:var(--text-color)">${data.account}</span>`;
		}

		// Bold total rows
		if (data.bold && column.fieldname !== "account") {
			value = `<strong>${value}</strong>`;
		}

		// Net Profit row: green (profit) or red (loss)
		if (column.fieldname !== "account") {
			let num = flt(data[column.fieldname]);

			if (data.is_net_profit) {
				let color = num >= 0 ? "var(--green-500)" : "var(--red-500)";

				value = `
					<span style="
						color:${color};
						font-weight:700;
					">
						${value}
					</span>
				`;
			} else if (data.root_type === "Income") {
				value = `
					<span style="color:var(--red-500)">
						${value}
					</span>
				`;
			} else if (data.root_type === "Expense") {
				value = `
					<span style="color:#3b82f6">
						${value}
					</span>
				`;
			}
		}

		return value;
	},

	// ─────────────────────────────────────────────────────────────
	// TREE OPTIONS
	// ─────────────────────────────────────────────────────────────
	tree: true,
	name_field: "account",
	parent_field: "parent_account",
	initial_depth: 2,

	get_datatable_options: function (options) {
		return Object.assign(options, {
			layout: "fixed",
			cellHeight: 36,
		});
	},

	// ── Show/hide exchange rate field and update its description ──
	_toggle_exchange_rate_field: function (selected_currency, company_currency) {
		let me = frappe.query_reports["Profit and Loss Statement"];
		let show = !!(
			selected_currency &&
			company_currency &&
			selected_currency !== company_currency
		);

		frappe.query_report.toggle_filter_display("presentation_exchange_rate", !show);

		if (!show) {
			frappe.query_report.set_filter_value("presentation_exchange_rate", 1);
			frappe.query_report.refresh(); // re-run with rate=1 (no conversion)
			return;
		}

		frappe.query_report.set_filter_value("presentation_exchange_rate", 1);
	},
};
